packages/foam-core/src/model/tags.ts

```diff
@@ -0,0 +1,183 @@
+import { FoamWorkspace, IDisposable, URI } from '../index';
+import { Resource } from './note';
+
+export type TagMetadata = { title: string; uri: URI };
+
+export class FoamTags implements IDisposable {
+  public readonly tags: { [key: string]: TagMetadata[] } = {};
+
+  constructor(private readonly workspace: FoamWorkspace) {}
+
+  /**
+   * List of disposables to destroy with the tags
+   */
+  private disposables: IDisposable[] = [];
+
+  /**
+   * Computes all the links in the workspace, connecting notes and
+   * creating placeholders.
+   *
+   * @param workspace the target workspace
+   * @param keepMonitoring whether to recompute the links when the workspace changes
+   * @returns the FoamGraph
+   */
+  public static fromWorkspace(
+    workspace: FoamWorkspace,
+    keepMonitoring: boolean = false
+  ): FoamTags {
+    let tags = new FoamTags(workspace);
+
+    Object.values(workspace.list()).forEach(resource =>
+      tags.resolveResource(resource)
+    );
+
+    if (keepMonitoring) {
+      tags.disposables.push(
+        workspace.onDidAdd(resource => {
+          tags.resolveResource(resource);
+        }),
+        workspace.onDidUpdate(change => {
+          tags.updateTagsForUpdatedResource(change.old, change.new);
+        }),
+        workspace.onDidDelete(resource => {
+          tags.updateTagsForRemovedResource(resource);
+        })
+      );
+    }
+    return tags;
+  }
+
+  updateTagsForUpdatedResource(oldVersion: Resource, newVersion: Resource) {
+    if (oldVersion.tags.size === 0 && newVersion.tags.size === 0) {
+      return;
+    }
+
+    this.resolveUpdatedTags(newVersion, oldVersion);
+    this.resolveUpdatedTitles(newVersion, oldVersion);
+    this.resolveMovedResource(newVersion, oldVersion);
+  }
+
+  updateTagsForRemovedResource(resource: Resource): void {
+    if (resource.tags.size === 0) {
+      return;
+    }
+
+    resource.tags.forEach(tag => {
+      this.tags[tag] = this.tags[tag].filter(
+        t => !URI.isEqual(t.uri, resource.uri)
+      );
+
+      if (this.tags[tag].length === 0) {
+        delete this.tags[tag];
+      }
+    });
+  }
+
+  resolveResource(resource: Resource): void {
+    resource.tags.forEach(tag => {
+      if (!this.tags[tag]) {
```

Comment:
minor: the pattern we use elsewhere is a bit different, here it is from `workspace.ts`:

```js
    this.resourcesByName[name] = this.resourcesByName[name] ?? [];
    this.resourcesByName[name].push(id);
```